

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Welcome to pipeline_dada2’s documentation! &mdash; pipeline_dada2 1.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="genindex.html"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="pipeline_dada2 1.0 documentation" href="#"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="#" class="icon icon-home"> pipeline_dada2
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <!-- Local TOC -->
              <div class="local-toc"><ul>
<li><a class="reference internal" href="#">Welcome to pipeline_dada2’s documentation!</a></li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a><ul>
<li><a class="reference internal" href="#run-dada2-amplicon-sequencing-analysis">run dada2 amplicon sequencing analysis</a><ul>
<li><a class="reference internal" href="#purpose">Purpose</a></li>
<li><a class="reference internal" href="#dependencies">Dependencies</a></li>
<li><a class="reference internal" href="#installation">Installation</a></li>
<li><a class="reference internal" href="#input-files">Input files</a></li>
<li><a class="reference internal" href="#parameterisation">Parameterisation</a></li>
<li><a class="reference internal" href="#output-files">Output files</a></li>
<li><a class="reference internal" href="#running-the-pipeline">Running the pipeline</a></li>
<li><a class="reference internal" href="#reporting">Reporting</a></li>
<li><a class="reference internal" href="#code">Code</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">pipeline_dada2</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="#">Docs</a> &raquo;</li>
        
      <li>Welcome to pipeline_dada2’s documentation!</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/index.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="welcome-to-pipeline-dada2-s-documentation">
<h1>Welcome to pipeline_dada2’s documentation!<a class="headerlink" href="#welcome-to-pipeline-dada2-s-documentation" title="Permalink to this headline">¶</a></h1>
<div class="toctree-wrapper compound">
</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></li>
<li><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></li>
<li><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></li>
</ul>
<span class="target" id="module-pipeline_dada2"></span><div class="section" id="run-dada2-amplicon-sequencing-analysis">
<h2>run dada2 amplicon sequencing analysis<a class="headerlink" href="#run-dada2-amplicon-sequencing-analysis" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Nick Ilott</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id$</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">Oct 26, 2018</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Python</td>
</tr>
</tbody>
</table>
<div class="section" id="purpose">
<h3>Purpose<a class="headerlink" href="#purpose" title="Permalink to this headline">¶</a></h3>
<p>The purpose of this pipeline is to take amplicon sequencing data and run it through
the DADA2 pipeline - this includes filtering and trimming reads, merging and clustering
reads and assigning reads to taxa.</p>
</div>
<div class="section" id="dependencies">
<h3>Dependencies<a class="headerlink" href="#dependencies" title="Permalink to this headline">¶</a></h3>
<p>In order to run this pipeline you will need to install the following packages:</p>
<ul class="simple">
<li>CGAT tools. The documentation can be found <a class="reference external" href="https://github.com/cgat-developers/cgat-flow">here</a></li>
</ul>
<ul class="simple">
<li>NGSKit. The code can be found in this <a class="reference external" href="https://github.com/nickilott/NGSKit">github repository</a>.</li>
</ul>
</div>
<div class="section" id="installation">
<h3>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h3>
<p>You should install CGAT tools using the installation script that is provided, this will
install everything you need into a conda environment.</p>
<p>To activate the enviroment do the following:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">source</span> <span class="o">&lt;/</span><span class="n">full</span><span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">folder</span><span class="o">/</span><span class="n">without</span><span class="o">/</span><span class="n">trailing</span><span class="o">/</span><span class="n">slash</span><span class="o">&gt;/</span><span class="n">conda</span><span class="o">-</span><span class="n">install</span><span class="o">/</span><span class="n">etc</span><span class="o">/</span><span class="n">profile</span><span class="o">.</span><span class="n">d</span><span class="o">/</span><span class="n">conda</span><span class="o">.</span><span class="n">sh</span>
<span class="n">conda</span> <span class="n">activate</span> <span class="n">base</span>
<span class="n">conda</span> <span class="n">activate</span> <span class="n">cgat</span><span class="o">-</span><span class="n">f</span>

<span class="c1"># finally, please run the cgatflow command-line tool to check the installation:</span>
<span class="n">cgatflow</span> <span class="o">--</span><span class="n">help</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">Make sure that you are installing CGAT into a clean environment i.e. with no modules
loaded from shared directories. You can do something like module purge to remove loaded
modules.</p>
</div>
<p>However it does not come with dada2 installed. Dada2 requires RcppParallel to be installed so
first, so once you have activated the CGAT environment install this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">conda</span><span class="o">-</span><span class="n">forge</span> <span class="n">r</span><span class="o">-</span><span class="n">rcppparallel</span> 
</pre></div>
</div>
<p>Then you should install the dada2 R package as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">bioconda</span> <span class="n">bioconductor</span><span class="o">-</span><span class="n">dada2</span>
</pre></div>
</div>
<p>This should install dada2 so that when you start R you can simply type:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">library</span><span class="p">(</span><span class="s2">&quot;dada2&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>You will in all likelihood also have to install optparse for argument parsing in the R scripts:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">conda</span><span class="o">-</span><span class="n">forge</span> <span class="n">r</span><span class="o">-</span><span class="n">optparse</span>
</pre></div>
</div>
<p>Once you have successfully installed CGAT, dada2 and optparse, you should then download the NGSKit
repository - this contains the dada2 pipeline (/pipelines/pipeline_dada2.py). In a development
folder run:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">git</span> <span class="n">clone</span> <span class="n">git</span><span class="nd">@github</span><span class="o">.</span><span class="n">com</span><span class="p">:</span><span class="n">nickilott</span><span class="o">/</span><span class="n">NGSKit</span><span class="o">.</span><span class="n">git</span>
</pre></div>
</div>
<p>It is possible that the modules that are required for the pipeline to run will have to be
placed into your PYTHONPATH:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>PYTHONPATH=$PYTHONPATH:&lt;path-to-NGSKit&gt;
export $PYTHONPATH
</pre></div>
</div>
</div>
<div class="section" id="input-files">
<h3>Input files<a class="headerlink" href="#input-files" title="Permalink to this headline">¶</a></h3>
<p>The input is a directory of fastq formatted files. These should be placed in the directory in
which you wish to run the pipeline. They must be of the format &lt;name&gt;.fastq.1.gz for single-end
data and two files &lt;name&gt;.fastq.1.gz and &lt;name&gt;.fastq.2.gz for paired-end data.</p>
</div>
<div class="section" id="parameterisation">
<h3>Parameterisation<a class="headerlink" href="#parameterisation" title="Permalink to this headline">¶</a></h3>
<p>There is a pipeline.yml file that specifies the parameters to be passed to dada2 and reporting
functions in the pipeline. This is located in &lt;path-to-NGSKit&gt;/pipelines/pipeline_dada2/pipeline.yml
and needs to be copied to your working directory. The parameters should be changed to suitable values.
and are explained in the dada2 <a class="reference external" href="https://benjjneb.github.io/dada2/tutorial.html">tutorial</a>.</p>
</div>
<div class="section" id="output-files">
<h3>Output files<a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h3>
<p>The main output file of the pipeline is the counts matrix that consists of amplicon sequence variants
and their abundance in each sample. The pipeline assigns taxonomy to each ASV and this is incorporated
into the ASV name in the resulting file. It is of the form:</p>
<table border="1" class="docutils">
<colgroup>
<col width="78%" />
<col width="10%" />
<col width="11%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>test_id</td>
<td>Sample1</td>
<td>Sample2</td>
</tr>
<tr class="row-even"><td>ASV1:p__phylum1;c__class1;o__order1;f__family1;g__genus1;s__species1</td>
<td>1000</td>
<td>1239</td>
</tr>
<tr class="row-odd"><td>ASV2:p__phylum2;c__class2;o__order2;f__family2;g__genus2;s__species2</td>
<td>500</td>
<td>10</td>
</tr>
<tr class="row-even"><td>ASV3:p__phylum3;c__class3;o__order3;f__family3;g__genus3;s__species3</td>
<td>1000</td>
<td>2300</td>
</tr>
</tbody>
</table>
<p>The purpose of this output file is that it can be taken forward in a easy fashion to look at differential
abundance using software such as DESeq2 and this will be done on a per ASV level. If you wish to perform
analysis on counts that have been summed over taxa at a particular taxonomic level you can use the following
output files:</p>
<ul class="simple">
<li>taxonomy_abundances.dir/phylum_abundances.tsv</li>
<li>taxonomy_abundances.dir/class_abundances.tsv</li>
<li>taxonomy_abundances.dir/order_abundances.tsv</li>
<li>taxonomy_abundances.dir/family_abundances.tsv</li>
<li>taxonomy_abundances.dir/genus_abundances.tsv</li>
<li>taxonomy_abundances.dir/species_abundances.tsv</li>
</ul>
</div>
<div class="section" id="running-the-pipeline">
<h3>Running the pipeline<a class="headerlink" href="#running-the-pipeline" title="Permalink to this headline">¶</a></h3>
<p>Running the pipeline should be as simple as:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">&lt;</span><span class="n">path</span><span class="o">-</span><span class="n">to</span><span class="o">-</span><span class="n">NGSKit</span><span class="o">&gt;/</span><span class="n">pipelines</span><span class="o">/</span><span class="n">pipeline_dada2</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">v5</span> <span class="o">-</span><span class="n">p50</span> <span class="n">make</span> <span class="n">full</span>
</pre></div>
</div>
<p>Set p to the number of processes that you want to run - this corresponds to the number of the samples
that you want to run (unless it is a huge number and exceeds a reasonable number to run on the cluster).</p>
</div>
<div class="section" id="reporting">
<h3>Reporting<a class="headerlink" href="#reporting" title="Permalink to this headline">¶</a></h3>
<p>Once the pipeline has finished you can build a report by running:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">&lt;</span><span class="n">path</span><span class="o">-</span><span class="n">to</span><span class="o">-</span><span class="n">NGSKit</span><span class="o">&gt;/</span><span class="n">pipelines</span><span class="o">/</span><span class="n">pipeline_dada2</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">v5</span> <span class="o">-</span><span class="n">p50</span> <span class="n">make</span> <span class="n">build_report</span>
</pre></div>
</div>
</div>
<div class="section" id="code">
<h3>Code<a class="headerlink" href="#code" title="Permalink to this headline">¶</a></h3>
<dl class="function">
<dt id="pipeline_dada2.addUniqueIdentifiers">
<code class="descclassname">pipeline_dada2.</code><code class="descname">addUniqueIdentifiers</code><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_dada2.addUniqueIdentifiers" title="Permalink to this definition">¶</a></dt>
<dd><p>add unique identifiers to track the sequences</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_dada2.assignTaxonomy">
<code class="descclassname">pipeline_dada2.</code><code class="descname">assignTaxonomy</code><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_dada2.assignTaxonomy" title="Permalink to this definition">¶</a></dt>
<dd><p>assign taxonomy to sequences</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_dada2.buildDefinitiveTable">
<code class="descclassname">pipeline_dada2.</code><code class="descname">buildDefinitiveTable</code><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_dada2.buildDefinitiveTable" title="Permalink to this definition">¶</a></dt>
<dd><p>build the final table with newids and
abundance information</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_dada2.build_report">
<code class="descclassname">pipeline_dada2.</code><code class="descname">build_report</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_dada2.build_report" title="Permalink to this definition">¶</a></dt>
<dd><p>render the rmarkdown report file</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_dada2.filterAndTrim">
<code class="descclassname">pipeline_dada2.</code><code class="descname">filterAndTrim</code><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_dada2.filterAndTrim" title="Permalink to this definition">¶</a></dt>
<dd><p>filter and trim reads based on user-defined
parameters</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_dada2.mergeAbundanceTables">
<code class="descclassname">pipeline_dada2.</code><code class="descname">mergeAbundanceTables</code><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_dada2.mergeAbundanceTables" title="Permalink to this definition">¶</a></dt>
<dd><p>combine sequence/abundance tables across
samples</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_dada2.mergeTaxonomyTables">
<code class="descclassname">pipeline_dada2.</code><code class="descname">mergeTaxonomyTables</code><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_dada2.mergeTaxonomyTables" title="Permalink to this definition">¶</a></dt>
<dd><p>combine sequence/taxonomy tables across
samples</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_dada2.runSampleInference">
<code class="descclassname">pipeline_dada2.</code><code class="descname">runSampleInference</code><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_dada2.runSampleInference" title="Permalink to this definition">¶</a></dt>
<dd><p>run dada2 sample inference. It goes through the following
stages:
1) learning error model
2) de-replication
3) merging reads (if paired)
4) removing chimeras</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_dada2.splitTableByTaxonomicLevels">
<code class="descclassname">pipeline_dada2.</code><code class="descname">splitTableByTaxonomicLevels</code><span class="sig-paren">(</span><em>infile</em>, <em>outfiles</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_dada2.splitTableByTaxonomicLevels" title="Permalink to this definition">¶</a></dt>
<dd><p>split the table at different taxonomic levels and
sum counts for each member of that level</p>
</dd></dl>

</div>
</div>
</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Nick Ilott.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>